### PR TITLE
Updating the new Integration Test Framework to properly support Dev15.

### PR DESCRIPTION
**Pull request by tannergooding**
FYI. @dotnet/roslyn-infrastructure

### CODE REVIEW

**Comment by tannergooding** [comment]
FYI. @jmarolf, since he had some concerns about integration tests in https://github.com/dotnet/roslyn/issues/13620

**Comment by tannergooding** [comment]
This also ports one more test: `CSharpAsyncOutput`.

I can split it out into it's own PR if desired.

**Comment by tannergooding** [comment]
test ci please

**Comment by tannergooding** [comment]
test ci please

The changes to Utilities.groovy to support the new image were not done yet

**Comment by tannergooding** [comment]
Ping. @jaredpar, @jasonmalinowski, @dotnet/roslyn-infrastructure

File: BuildAndTest.proj [hunk 1]
```diff
@@ -13,17 +13,26 @@
     <Configuration Condition="'$(Configuration)' == ''">Debug</Configuration>
     <RunTestArgs Condition="'$(ManualTest)' == ''">$(RunTestArgs) -xml</RunTestArgs>
     <RunTestArgs Condition="'$(Test64)' == 'true'">$(RunTestArgs) -test64</RunTestArgs>
+    <RunTestArgs Condition="'$(TestVSI)' == 'true'">$(RunTestArgs) -testVSI</RunTestArgs>
     <RunTestArgs Condition="'$(Trait)' != ''">$(RunTestArgs) -trait:$(Trait)</RunTestArgs>
     <RunTestArgs Condition="'$(NoTrait)' != ''">$(RunTestArgs) -notrait:$(NoTrait)</RunTestArgs>
-    <IncludePattern Condition="'$(IncludePattern)' == ''">*.UnitTests*.dll</IncludePattern>
+    <IncludePattern Condition="'$(IncludePattern)' == '' AND '$(TestVSI)' != 'true'">*.UnitTests*.dll</IncludePattern>
+    <IncludePattern Condition="'$(IncludePattern)' == '' AND '$(TestVSI)' == 'true'">*.IntegrationTests*.dll</IncludePattern>
```
Comment (jaredpar) [inline BuildAndTest.proj]: > <IncludePattern Condition="'$(IncludePattern) [](start = 3, length = 46)

These lines mean I can't test VSI and non-VSI in the same run correct?

File: src/Tools/Source/RunTests/ITestExecutor.cs [hunk 2]
```diff
@@ -16,13 +16,16 @@ internal struct TestExecutionOptions
         internal bool UseHtml { get; }
         internal bool Test64 { get; }
 
-        internal TestExecutionOptions(string xunitPath, string trait, string noTrait, bool useHtml, bool test64)
+        internal bool TestVSI { get; }
```
Comment (jaredpar) [inline src/Tools/Source/RunTests/ITestExecutor.cs]: >  internal bool TestVSI { get; } [](start = 6, length = 32)

Please maintain the existing grouping of no spaces between the proprties

File: BuildAndTest.proj [hunk 3]
```diff
@@ -13,17 +13,26 @@
     <Configuration Condition="'$(Configuration)' == ''">Debug</Configuration>
     <RunTestArgs Condition="'$(ManualTest)' == ''">$(RunTestArgs) -xml</RunTestArgs>
     <RunTestArgs Condition="'$(Test64)' == 'true'">$(RunTestArgs) -test64</RunTestArgs>
+    <RunTestArgs Condition="'$(TestVSI)' == 'true'">$(RunTestArgs) -testVSI</RunTestArgs>
```
Comment (jaredpar) [inline BuildAndTest.proj]: >VSI [](start = 35, length = 3)

Why VSI instead of Vsi?  Everywhere else we use Html instead of HTML

File: netci.groovy [hunk 4]
```diff
@@ -198,6 +198,27 @@ commitPullList.each { isPr ->
...
+commitPullList.each { isPr ->
+  def jobName = Utilities.getFullJobName(projectName, "open-vsi", isPr)
+  def myJob = job(jobName) {
+    description('open integration tests')
+    label('auto-win2012-20160912')
+    steps {
+      batchFile("""set TEMP=%WORKSPACE%\\Binaries\\Temp
+mkdir %TEMP%
+set TMP=%TEMP%
+set VS150COMNTOOLS=%ProgramFiles(x86)%\\Microsoft Visual Studio\\VS15Preview\\Common7\\Tools\\
```
Comment (jaredpar) [inline netci.groovy]: >VS150COMNTOOLS [](start = 4, length = 14)

Why are we overriding VS environment variables?

File: netci.groovy [hunk 5]
```diff
@@ -198,6 +198,27 @@ commitPullList.each { isPr ->
...
+// Open Integration Tests
+commitPullList.each { isPr ->
+  def jobName = Utilities.getFullJobName(projectName, "open-vsi", isPr)
+  def myJob = job(jobName) {
+    description('open integration tests')
+    label('auto-win2012-20160912')
+    steps {
+      batchFile("""set TEMP=%WORKSPACE%\\Binaries\\Temp
+mkdir %TEMP%
+set TMP=%TEMP%
```
Comment (jaredpar) [inline netci.groovy]: Inlining this temp code everywhere makes my eyes bleed.  We really need to find a better way to do this.  Not really the fault of this change though.

**Comment by jaredpar** [comment]
>     internal struct TestExecutionOptions

Given the change in the current form you will end up partipating in test caching.  I don't think that is what you want here for VSI tests.  Caching should probably be disabled for them. 

---
Refers to: src/Tools/Source/RunTests/ITestExecutor.cs:11 in dc8c01c. [](commit_id = dc8c01c7db238a3b94918237ec26ec7e9245b428, deletion_comment = False)

File: src/VisualStudio/Core/Def/ServicesVisualStudio.csproj [hunk 6]
```diff
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" TreatAsLocalProperty="DeployExtension">
```
Comment (jaredpar) [inline src/VisualStudio/Core/Def/ServicesVisualStudio.csproj]: >TreatAsLocalProperty="DeployExtension"> [](start = 111, length = 39)

Why do we need this?

File: src/VisualStudio/IntegrationTests/AbstractEditorTest.cs [hunk 7]
```diff
@@ -11,16 +11,14 @@
...
 namespace Roslyn.VisualStudio.IntegrationTests
 {
-    public abstract class AbstractEditorTests : IDisposable
+    public abstract class AbstractEditorTest : AbstractIntegrationTest
     {
-        private readonly VisualStudioInstanceContext _visualStudio;
-        private readonly VisualStudioWorkspace_OutOfProc _visualStudioWorkspaceOutOfProc;
-        private readonly Editor_OutOfProc _editor;
+        protected readonly VisualStudioWorkspace_OutOfProc _visualStudioWorkspaceOutOfProc;
+        protected readonly Editor_OutOfProc _editor;
```
Comment (jaredpar) [inline src/VisualStudio/IntegrationTests/AbstractEditorTest.cs]: >_editor [](start = 44, length = 7)

Protected variables should be pascal cased.

File: src/VisualStudio/IntegrationTests/AbstractIntegrationTest.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+using System;
+using System.Threading;
+using Roslyn.VisualStudio.Test.Utilities;
+
+namespace Roslyn.VisualStudio.IntegrationTests
+{
+    public abstract class AbstractIntegrationTest : IDisposable
+    {
+        protected readonly VisualStudioInstanceContext _visualStudio;
```
Comment (jaredpar) [inline src/VisualStudio/IntegrationTests/AbstractIntegrationTest.cs]: >_visualStudio; [](start = 55, length = 14)

Protected variables should be pascal cased.

File: src/VisualStudio/IntegrationTests/AbstractInteractiveWindowTest.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+namespace Roslyn.VisualStudio.IntegrationTests
+{
+    public abstract class AbstractInteractiveWindowTest : AbstractIntegrationTest
+    {
+        private const string Edit_SelectionCancelCommand = "Edit.SelectionCancel";
+
+        private static readonly char[] LineSeparators = { '\r', '\n' };
+
+        protected readonly CSharpInteractiveWindow_OutOfProc _interactiveWindow;
```
Comment (jaredpar) [inline src/VisualStudio/IntegrationTests/AbstractInteractiveWindowTest.cs]: >_interactiveWindow [](start = 61, length = 18)

Protected vnarciables should be pascal cased.  

Going to stop adding this comment.  Please just assume it for every field.

File: src/VisualStudio/IntegrationTests/AbstractInteractiveWindowTest.cs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+using Roslyn.VisualStudio.Test.Utilities.OutOfProcess;
+using Xunit;
+
+namespace Roslyn.VisualStudio.IntegrationTests
+{
+    public abstract class AbstractInteractiveWindowTest : AbstractIntegrationTest
+    {
+        private const string Edit_SelectionCancelCommand = "Edit.SelectionCancel";
+
+        private static readonly char[] LineSeparators = { '\r', '\n' };
```
Comment (jaredpar) [inline src/VisualStudio/IntegrationTests/AbstractInteractiveWindowTest.cs]: >private static readonly char[] LineSeparators = { '\r', '\n' }; [](start = 8, length = 63)

This is not the complete set of line separators that Visual Studio recognizes.  In that okay for this test?

File: src/VisualStudio/IntegrationTests/SharedIntegrationHostFixture.cs [hunk 11]
```diff
@@ -8,6 +9,11 @@ namespace Roslyn.VisualStudio.IntegrationTests
     [CollectionDefinition(nameof(SharedIntegrationHostFixture))]
     public sealed class SharedIntegrationHostFixture : ICollectionFixture<VisualStudioInstanceFactory>
     {
+        public const string RoslynLanguageServicesPackageId = "0b5e8ddb-f12d-4131-a71d-77acc26a798f";
```
Comment (jaredpar) [inline src/VisualStudio/IntegrationTests/SharedIntegrationHostFixture.cs]: >"0b5e8ddb-f12d-4131-a71d-77acc26a798f"; [](start = 62, length = 39)

This should be a constant that we reference from elsewhere.

File: BuildAndTest.proj (hunk 1 above)
Comment (tannergooding) [inline BuildAndTest.proj]: Correct. Same as you can't test 32 and 64 in the same run. We can definitely change it (eventually) so that unit and VSI can run in parallel.

File: netci.groovy (hunk 4 above)
Comment (tannergooding) [inline netci.groovy]: Because the VS environment variable doesn't exist in Dev15 Preview 4 and we haven't implemented the general build level logic to detect where VS is installed yet.

File: netci.groovy (hunk 5 above)
Comment (tannergooding) [inline netci.groovy]: We could put it in a bat file that this calls, but I'd rather do that in a separate PR.

File: src/VisualStudio/Core/Def/ServicesVisualStudio.csproj (hunk 6 above)
Comment (tannergooding) [inline src/VisualStudio/Core/Def/ServicesVisualStudio.csproj]: So that the explicit `DeployExtension=false` required for this project (since it doesn't include a source extension manifest file) takes precedence over any user defined option.

File: src/VisualStudio/IntegrationTests/AbstractInteractiveWindowTest.cs (hunk 10 above)
Comment (tannergooding) [inline src/VisualStudio/IntegrationTests/AbstractInteractiveWindowTest.cs]: For this test (and all of the Interactive Window Tests we have) yes.

File: src/VisualStudio/IntegrationTests/SharedIntegrationHostFixture.cs (hunk 11 above)
Comment (tannergooding) [inline src/VisualStudio/IntegrationTests/SharedIntegrationHostFixture.cs]: This will actually need to change, the packages aren't the VSIX ID (like I assumed) it is the VS Authoring name

File: BuildAndTest.proj [hunk 12]
```diff
@@ -102,7 +111,7 @@
       <RunTestsArgs>$(NuGetPackageRoot)\xunit.runner.console\$(xunitrunnerconsoleVersion)\tools $(RunTestArgs) @(TestAssemblies, ' ')</RunTestsArgs>
     </PropertyGroup>
     
-    <Exec Condition="'$(RunProcessWatchdog)' != 'true'" Command="&quot;$(CoreRunExe)&quot; $(CoreRunArgs)" />
+    <Exec Condition="'$(RunProcessWatchdog)' != 'true' AND '$(TestVSI)' != 'true'" Command="&quot;$(CoreRunExe)&quot; $(CoreRunArgs)" />
```
Comment (jaredpar) [inline BuildAndTest.proj]: > <Exec Condition="'$(RunProcessWatchdog)' ! [](start = 3, length = 43)

Did we ever actually enable process watch dog?  I don't think we did.

File: netci.groovy (hunk 5 above)
Comment (jaredpar) [inline netci.groovy]: Yes definitely a separate Pr. 

---
In reply to: [78596261](https://github.com/dotnet/roslyn/pull/13732#discussion_r78596261) [](ancestors = 78596261)

File: BuildAndTest.proj (hunk 12 above)
Comment (tannergooding) [inline BuildAndTest.proj]: I don't think so either.

File: src/Tools/Source/RunTests/ITestExecutor.cs (hunk 2 above)
Comment (tannergooding) [inline src/Tools/Source/RunTests/ITestExecutor.cs]: Fixed.

File: BuildAndTest.proj (hunk 3 above)
Comment (tannergooding) [inline BuildAndTest.proj]: Oversight on my part. fixing.

File: src/VisualStudio/IntegrationTests/AbstractIntegrationTest.cs (hunk 8 above)
Comment (tannergooding) [inline src/VisualStudio/IntegrationTests/AbstractIntegrationTest.cs]: Fixed.

File: src/VisualStudio/IntegrationTests/AbstractInteractiveWindowTest.cs (hunk 9 above)
Comment (tannergooding) [inline src/VisualStudio/IntegrationTests/AbstractInteractiveWindowTest.cs]: Fixed.

**Comment by tannergooding** [comment]
@jaredpar, what needs to be done to disable caching?

File: src/Tools/Source/RunTests/Options.cs [hunk 13]
```diff
@@ -77,6 +82,11 @@ internal static Options Parse(string[] args)
                     opt.Test64 = true;
                     index++;
                 }
+                else if (comp.Equals(current, "-testVsi"))
+                {
+                    opt.TestVsi = true;
```
Comment (jaredpar) [inline src/Tools/Source/RunTests/Options.cs]: Add the following

```
opt.UseCachedResults = false;
```

Comment (tannergooding) [inline src/Tools/Source/RunTests/Options.cs]: Fixed.

**Comment by tannergooding** [comment]
@jaredpar, caching is disabled. Anything else needed?

**Comment by tannergooding** [comment]
ping @jaredpar, the open-vsi changes are needed for another repo as well, so I would appreciate sign-off if possible (now that I believe all the issues you brought up are resolved).

Additionally, we really need to move the open-vsi framework itself to its own repo 😄. That repo would contain the base level functionality for VS in general. Roslyn can then provide its own binary that allows writing unit tests against Roslyn specific features (and other repo's can do the same).

**Comment by tannergooding** [comment]
FYI. @jaredpar, updated the comment to explain why we only want one integration test to run at a time. Will merge after the jobs are passing again.
